Optimize log print by using html format

woehrl01 opened the pull request:
See facebook/yoga#453. Optimizes the node log print by generating some enum text via ```enum.py``` and moving printing to new functions to reduce boilerplate code. 

Changes the log output to format the nodes in html to be able to copy paste it  into browsers for quick debugging. 

Hides all default values.

emilsjolander commented on yoga/YGEnumsPrint.h:
```diff
@@ -0,0 +1,359 @@
+
+YG_EXTERN_C_BEGIN
+
+void YGLogAlign(YGLogLevel logLevel, const char * param, YGAlign value){
```
What is `param` ? Could you choose a more descriptive name?

emilsjolander replied:
Also could you mark all parameters as const?

emilsjolander replied:
File is names `YGEnumsPrint` but functions are named `YGLogX`. The mix of `Print` and `Log` is a bit odd.

emilsjolander commented on yoga/YGEnumsPrint.h:
```diff
@@ -0,0 +1,359 @@
+      break;
+  }
+  if(param == NULL) {
+    YGLog(logLevel, text);
```
I would rather the functions in this file not return log but instead return a statically allocated string which is logged by the caller.

woehrl01 replied:
I also thought of this, but this makes the callside a bit more boilerplate, not sure if we want this.

emilsjolander replied:
Could you give it a try? I think it might be worth it. You can always abstract if if(param == null) part into another function.

emilsjolander commented on enums.py:
```diff
@@ -146,6 +156,34 @@ def to_java_upper(symbol):
     f.write('YG_EXTERN_C_END\n')
 
+# write out C headers for printing
+with open(root + '/yoga/YGEnumsPrint.h', 'w') as f:
```
Could this just be part of YGEnums.h? Collocating this stuff is usually pretty nice

emilsjolander commented on yoga/Yoga.c:
```diff
@@ -781,14 +772,46 @@ static void YGPrintNumberIfNotUndefinedf(const char *str, const float number) {
+  \
+}
+
 static bool YGFourValuesEqual(const YGValue four[4]) {
```
This function doesn't even handle start & end or vertical & horizontal so would love it fixes for that were included here as well 👍

emilsjolander commented on yoga/Yoga.c:
```diff
@@ -781,14 +772,46 @@ static void YGPrintNumberIfNotUndefinedf(const char *str, const float number) {
+YGPrintNumberIfNotZero(YG_STRING(type) YG_STRING(edge), \
+                       YGComputedEdgeValue(node->style.type, YGEdge##edge, &YGValueZero));
+
+#define YG_LOG_PRINT_NONZERO_EDGES(type)                                                     \
```
This could probably be a function instead of a macro which takes a array of YGValues

woehrl01 replied:
this only works nicely if we write the position in lowercase, example: ```margin-left``` instead of current ```marginLeft```. Could be nicer if we write html instead (see suggestion below)

emilsjolander replied:
Not sure I fully understand

woehrl01 replied:
If we loop over the array we could use ```YGLogEdge``` to stringify ```YGEdge``` but that returns only the edge in lowercase. So we need some fiddeling to uppercase if we want to stick to something like ```marginLeft```. 

But as you think that an html layout would be a great improvment, this is more easily doable (no need to uppercase). I'll update the PR with your suggestions and we can discuss on top of that, what do you think?

emilsjolander replied:
sounds good!

woehrl01 commented:
Thanks for the comments, I'll try to work these in.

Additonally I'm thinking if it would be more useful if we output those values as html instead of JSONlike, to be able do copy paste possible wrong layouts more easily. 

But I don't know how the output is used currently (e.g. ReactNative).

emilsjolander commented:
Outputting a html layout would be amazing. This isn't really used a lot so I think outputting html will make it much more useful. Love the idea.

woehrl01 commented:
I added html output and removed the usage for a macro, there are some small things which aren't that nice yet (e.g. ```""``` usage). Those will be removed later, when I find a better way for this.

> This function doesn't even vertical & horizontal ...

I can't add those as accessing this values lead to an ```YG_ASSERT``` error.

> Could this just be part of YGEnums.h? Collocating this stuff is usually pretty nice

not really as we get some linker errors later, we would need to add the function templates to the header and provide an additional ```*.c``` file. Should we go this way?

> File is names YGEnumsPrint but functions are named YGLogX

I change this if we want to keep it that way, depends on the previous.

woehrl01 commented:
I'm not sure if we should print ```pt```, if we want to use the output as input for gentest, we should print ```px```, do not get different outputs based on those different scale factors. What do you think?

emilsjolander commented:
> we would need to add the function templates to the header and provide an additional *.c file. Should we go this way?

Yeah i'm totally ok with a `YGEnums.c` file.

> I'm not sure if we should print pt, if we want to use the output as input for gentest, we should print px, to not get different outputs based on those different scale factors. What do you think?

perhaps use px if scale factor is 1 otherwise pt? 

> I can't add those as accessing this values lead to an YG_ASSERT error.

Actually we probably should not be calling `YGComputedEdgeValue()` here are we want to print the value which was set on the style, not the computed value.

woehrl01 commented:
> perhaps use px if scale factor is 1 otherwise pt?

isn't css using different factors for pt and px? I'd prefere sticking with just px.

> print the value which was set on the style, not the computed value.

correct, I'll change it accordingly.

emilsjolander commented:
> isn't css using different factors for pt and px? I'd prefere sticking with just px.

 Ah yeah that sounds good.

emilsjolander commented:
@woehrl01 not sure. Also not sure why there is no output in the logs. Can you run the tests locally?

emilsjolander commented on tests/YGLoggerTest.cpp:
```diff
@@ -0,0 +1,63 @@
+#include <yoga/Yoga.h>
+
+static char writeBuffer[4096];
+static int _unmanagedLogger(YGLogLevel level, const char *format, va_list args) {
```
va_list is defined in stdarg.h right?

woehrl01 replied:
should I include this header?

emilsjolander replied:
yeah 👍

emilsjolander commented on yoga/YGEnums.c:
```diff
@@ -0,0 +1,234 @@
+
+#include "YGEnums.h"
+
+const char * YGStringifyAlign(const YGAlign value){
```
`YGAlignToString`

emilsjolander replied:
Also `char *YGStringifyAlign` instead of `char * YGStringifyAlign` (asterix placement)

emilsjolander commented on yoga/YGEnums.c:
```diff
@@ -0,0 +1,234 @@
+      return "space-between";
+    case YGAlignSpaceAround:
+      return "space-around";
+    default:
```
Is the default needed? I think this switch should be exhaustive. Not supplying default will allow the compiler to warn when not all cases are specified.

woehrl01 replied:
not exactly as a default. but at least as a "fall-through". Will remove the default and put the return "unknown" after the switch

emilsjolander replied:
What is the benefit of this?

woehrl01 replied:
the compiler needs it, the function doesn't have a return statement otherwise for any call case. (e.g. calling it an invalid argument)

emilsjolander commented on yoga/YGEnums.h:
```diff
@@ -132,4 +132,20 @@ typedef YG_ENUM_BEGIN(YGWrap) {
 } YG_ENUM_END(YGWrap);
 
+
+WIN_EXPORT const char * YGStringifyAlign(const YGAlign value);
```
`char *YG...`

emilsjolander replied:
Can you put each toString function bellow the corresponding enum declaration?

woehrl01 replied:
will do

woehrl01 commented:
Current output looks like:

```html
<div layout="width: 0; height: 0; top: 0; left: 0;" style="flex-direction: column; justify-content: flex-start; align-items: stretch; align-content: flex-start; flex-grow: 0; flex-shrink: 0; " >
  <div layout="width: 0; height: 0; top: 0; left: 0;" style="flex-direction: column; justify-content: flex-start; align-items: stretch; align-content: flex-start; flex-grow: 0; flex-shrink: 0; " ></div>
  <div layout="width: 0; height: 0; top: 0; left: 0;" style="flex-direction: column; justify-content: flex-start; align-items: stretch; align-content: flex-start; flex-grow: 0; flex-shrink: 0; " ></div>
</div>
```

emilsjolander commented:
> Current output looks like:

Lovely!

woehrl01 commented:
should I hide all default values e.g. for ```flex-direction```, ```justify-content```, too?

What about ```flex``` ?

emilsjolander commented:
> should I hide all default values

Good point. I think hiding defaults is good.

> What about flex

Let's print it if it is set. negative or not.

woehrl01 commented:
Found the root cause of the failing travis test. All DEATH tests should be named like ```*DeathTest``` so they are executed at first, as they are run l later, they output seems to be overridden by our new logger tests. See: https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md

arcanis commented:
The JSON format had nice properties (for example, extracting a tree from an application and recreating it standalone via the JS bindings in order to better debug the library). Is there plans to support multiple logging outputs?

woehrl01 commented:
Multiple logging outputs could be doable. Did you use the JSONlike output? As the output wasn't exactly JSON ```'``` instead of ```"``` for properties. Even so it could be nice to use the output directly for javascript. I think the HTML output is more generic as you could use ```gentest.py``` to generate your DOM easily for any supported language.

emilsjolander commented:
@woehrl01 I agree we should stick with html output for now.
@arcanis I'm definetly not against multiple output formats but for now html is more useful
